lib/utils/lifecycle.js

```diff
@@ -75,18 +76,38 @@ function lifecycle (pkg, stage, wd, unsafe, failOk, cb) {
     env.npm_node_execpath = env.NODE = env.NODE || process.execPath
     env.npm_execpath = require.main.filename
 
-    // 'nobody' typically doesn't have permission to write to /tmp
-    // even if it's never used, sh freaks out.
-    if (!npm.config.get('unsafe-perm')) env.TMPDIR = wd
+    _getTemporaryDirectory(wd, process.env, function (er, tmpdir) {
+      if (er) return cb(er)
 
-    lifecycle_(pkg, stage, wd, env, unsafe, failOk, cb)
+      env.TMPDIR = tmpdir
+      lifecycle_(pkg, stage, wd, env, unsafe, failOk, cb)
+    })
   })
 }
 
 function _incorrectWorkingDirectory (wd, pkg) {
   return wd.lastIndexOf(pkg.name) !== wd.length - pkg.name.length
 }
 
+function _getTemporaryDirectory (wd, env, cb) {
+  if (_isWritable(env.TMPDIR)) return cb(null, env.TMPDIR)
+
+  // 'nobody' typically doesn't have permission to write to /tmp
+  // even if it's never used, sh freaks out.
+  if (_isWritable(wd)) return cb(null, wd)
+
+  cb(new Error('Cannot find writable TMPDIR, please set TMPDIR to something writable'))
+}
+
+function _isWritable (dir) {
+  try {
+    fs.accessSync(dir, fs.W_OK)
+    return true
+  } catch (e) {
+    return false
+  }
```

Comment:
I'm not a fan of using exceptions in this way, but it felt like an awkward place to have to use callbacks. If you'd prefer that let me know and I can switch it out.